### ISSUE TITLE
Feature Request : Sass mixins

### ISSUE THREAD

**Issue by froger** [Dogfalo/materialize#2869]
Hello, 
I wanted to include materialize into a current project, and we use a BEM naming, trying to keep Html for markup, CSS for styling. 
The problem I have with the current implementation, is that I need to use HTML markup to make styles. 

Here what I was thinking doing for example : 

```
<div class="form__field--active">
   <label class="form__label" for="UserEmail">
    <input type="text" name="user[email]" value="email" />
</div>
<div class="form__field">
   <label class="form__label" for="UserFirstname">
    <input type="text" name="user[firstname]" value="" />
</div>
```

and in the scss : 

```
@import "materialize/mixins/forms";
.form__field{
   @include input-field();
}

.form__field--active{
   @include input-field();
   @include input-field-active();
}
```

This way, I could keep clean html markup !
Here an example of mixins I was thinking (untested & using foundation) : 

``````
/** 
 * Define an input field (group of label + input).
 * When the input have a value, should see `input-field-active`.
 * Example : 
 * ```
 * .my-div{
 *   @include input-field(); // default format, see your variables
 * }
 * .my-div2{
 *   @include input-field($margin-top: 18px); // Override the top margin.
 * }
 * ```
 */
@mixin input-field(
  $margin-top: 1rem, 
  $font-size: 1em,
  $gutter-width: $gutter-width // use default materialize variable, and can be overriden
){
  $font-size: -zf-bp-to-em($font-size);
  margin-top: $margin-top;
  position: relative;
  & label{
    @include label(
      $label-color: $input-border-color,
      $font-size: rem-calc($font-size)
    );
    position: absolute;
    top: rem-calc($font-size - 2/10);
    left: $gutter-width / 2;
    cursor: text;
    transition: .2s ease-out;
  }
}

/**
 * Input field that have a value setted.
 * Example : 
 * ```
 * .my-div{
 *   @include input-field($font-size: 1.2rem);
 *   @include input-field-active();
 * }
 * ```
 */
@mixin input-field-active(){
  transform: translateY(-140%);
}

/**
 * Set the global HTML markup for the whole project.
 * For the ones that want to keep using materialize classes.
 */
@mixin global-input(){
  .input-field{
    @include input-field();
  }
  .input-field.active{
    @include input-field();
    @include input-field-active();
  }
}
``````

In summary, my feature request is to use a bit more Sass features in the whole project. All the defined classes should use _only_ mixins. And there is also the possibility here to stop adding classes that we don't use. 
Will execute CSS faster, make a smaller file, and be able to work with foundation or bootstrap.
Materialize should be super easy to customize, and this is actually not really the case.

Thanks for reading my bad english, and um abraço do Brasil.

**Comment by fega** [comment]
There is a great way to avoid the tech-deb.
but what about the js? It seems to be a refactor problem, and could cause a lot of issues, I thought.

off-topic: I like to use grammarly.com to improve my writing, it's also used also by native speakers!

Un abrazo desde Colombia.

**Comment by froger** [comment]
Well you are right ! After looking closer to the sources, it's looks like the classenames are not in options, but hard-coded in the js. I tought as everything is jquery plugin in materialize you would have the ability to specify your classes when instanciate them.

From the start it wasn't planned this way, and it seems difficult to make any moves without breaking everything. Seeing no movement for the past year, I would suggest you to close this feature request.

Thanks
